src/java/org/apache/cassandra/repair/consistent/LocalSessions.java

```diff
@@ -862,6 +860,25 @@ public void onFailure(Throwable t)
         }, MoreExecutors.directExecutor());
     }
 
+    /**
+     * Checks for the session state, and sets it to prepared unless it is on a failed state.
+     * Making the checks inside a synchronized block to prevent the session state from
+     * being changed between the read and the update.
+     *
+     * @param session The local session to be set to prepared.
+     * @return true if the session is prepared, false if not, i.e. session failed
+     */
+    private boolean prepareSessionExceptFailed(LocalSession session) {
```

Comment:
wdyt about overloading `setStateAndSave()` to have compare and set semantics? That would be a generalization of this one that could be more useful in the future.

Reply:
What do you suggest that should be compared inside the `setStateAndSave()` method? Would it be a fallback of the `Preconditions.checkArgument` checking the `canTransitionTo` of a given state, so instead of failing the method execution we just swallow it silently and log it?

If that's the suggestion, I see that the method is being used in several other places:
* handleFinalizeProposeMessage
* handleFinalizeCommitMessage
* handleStatusResponse
* cancelSession
* maybeSetRepairing

And not all of them have follow a strict compare and set pattern (some of them just have a set).

Having said that, if we wanted to overload setStateAndSave, my opinion is that it would be a slightly out of the scope change that may somehow force us to add fallbacks where we have currently none (those methods which only have a set). 

I think it may require a bit of a broader discussion to basically answer this question: Are we sure we want to avoid failing fast on those cases? I am definitely open to debate.

Reply:
The suggestion is to add to:
-  `setStateAndSave()` and argument `currentState` that can be null. If null the method behaves as it does now. If a state is provided it will then compare `state.getState()` to `currentState` and fail or return false if they don't match. Rename the method to `compareAndSaveState`i.e.
- Add a new `setStateAndSave`method without the `currentState` argument for all calls that don't need the compare behavior.

Reply:
@bereng , understanding the motivation to generalize the CAS-like operation, I'd like to point out the logic here is mostly, if _not_ in some state, then set and save. There is also a call-site that compares with 2 states. 
```java
        if (response.state == FINALIZED || response.state == FAILED)
        {
            setStateAndSave(session, response.state);
            logger.info("Unfinished local incremental repair session {} set to state {}", sessionID, response.state);
        }
```

Reply:
Well if you guys don't like it I'm fine with it. It's just how I would do it but again, personal opinion, no big thing. +1